### ISSUE TITLE
Add more control types. Combine controls with `and`.

### ISSUE THREAD

**Issue by Strilanc** [ProjectQ-Framework/ProjectQ#13]
These constructs save several boilerplate in the grover example. For example:

```python
def run_grover(eng, n, oracle):
	x = eng.allocate_qureg(n)
	All(H) | x
	
	with Loop(eng, int(math.pi/4.*math.sqrt(1 << n))):
		oracle(eng, x)
		with PlusControl(eng, x):
			eng.PhaseFlip()
	
	Measure | x
	eng.flush()
	
	return [int(qubit) for qubit in x]


def alternating_bits_oracle(eng, qubits):
	with Control(eng, qubits[1::2]) and OffControl(eng, qubits[::2]):
		eng.PhaseFlip()
```

An `OffControl` requires that the qubit be |0>. It's equivalent to X-before, Control-during, X-after.

A `PlusControl` requires that the state be |0>+|1>. It's equivalent to H,X-before, Control-during, X,H-after.

**Comment by thomashaener** [comment]
Yes, we could start with extending ``Control`` s.t. it takes an additional parameter (which is ``AllOnes`` by default), e.g.,
```
with Control(eng, x, predicate=AllZero):
  ...
```
Control would then apply the predicate before and after the controlled operation(s).

We can then provide specializations such as the ones you mentioned.
I also like combining them with ``and``.

**Comment by Takishima** [comment]
Closed by #400 .
